3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -248,97 +339,134 @@ object PiiPseudonymizerEnrichment extends ParseableEnrichment {
+      (compact(render(parse(fieldValue) match {
+        case JObject(jObject) =>
           val jObjectMap = jObject.toMap
           val updated = jObjectMap.filterKeys(_ == "data").mapValues {
             case JArray(contexts) =>
               JArray(contexts.map {
-                case JObject(context) => modifyObjectIfSchemaMatches(context)
-                case x                => x
+                case JObject(context) =>
+                  val (values, listOfModifiedValues) = modifyObjectIfSchemaMatches(context, strategy)
+                  modifiedFields ++= listOfModifiedValues
+                  values
+                case x => x
               })
-            case JObject(unstructEvent) => modifyObjectIfSchemaMatches(unstructEvent)
-            case x                      => x
+            case JObject(unstructEvent) =>
+              val (values, listOfModifiedValues) = modifyObjectIfSchemaMatches(unstructEvent, strategy)
+              modifiedFields ++= listOfModifiedValues
+              values
+            case x => x
           }
           JObject((jObjectMap ++ updated).toList)
-        }
         case x => x
-      }))
-    } else null
+      })), modifiedFields.toList)
+    } else (null, modifiedFields.toList)
+  }
 
-  private def modifyObjectIfSchemaMatches(context: List[(String, json4s.JValue)]): JObject = {
+  /**
+   * Tests whether the schema for this event matches the schema criterion and if it does modifies it.
+   */
+  private def modifyObjectIfSchemaMatches(context: List[(String, json4s.JValue)],
+                                          strategy: PiiStrategy): (JObject, List[JsonModifiedField]) = {
     val fieldsObj = context.toMap
     (for {
-      schema              <- fieldsObj.get("schema")
-      parsedSchemaMatches <- SchemaKey.parse(schema.extract[String]).map(schemaCriterion.matches).toOption
+      schema <- fieldsObj.get("schema")
+      schemaStr = schema.extract[String]
+      parsedSchemaMatches <- SchemaKey.parse(schemaStr).map(schemaCriterion.matches).toOption
       data                <- fieldsObj.get("data")
       if parsedSchemaMatches
-    } yield JObject(fieldsObj.updated("schema", schema).updated("data", jsonPathReplace(data)).toList)).getOrElse(JObject(context))
+      updated = jsonPathReplace(data, strategy, schemaStr)
+    } yield
+      (JObject(fieldsObj.updated("schema", schema).updated("data", updated._1).toList), updated._2)).getOrElse((JObject(context), List()))
   }
 
   // Configuration for JsonPath
   private val JsonPathConf =
-    Configuration.builder().options(JOption.SUPPRESS_EXCEPTIONS).jsonProvider(new JacksonJsonNodeJsonProvider()).build()
+    Configuration
+      .builder()
+      .options(JOption.SUPPRESS_EXCEPTIONS, JOption.ALWAYS_RETURN_LIST)
+      .jsonProvider(new JacksonJsonNodeJsonProvider())
+      .build()
 
   /**
    * Replaces a value in the given context data with the result of applying the strategy that value.
```

Comment:
this comment might need refreshing given the function seems to have changed significantly